### ISSUE TITLE
Outcome reward model code + chapter

### ISSUE THREAD

**Issue by LazaroHurtado** [natolambert/rlhf-book#505]
I think there are some issues in the definition of ORM in chapter 5 and the training script. The ORM section of chapter 5 says:

> Formally, following [[11]](https://rlhfbook.com/c/05-reward-models#ref-lyu2025exploring) this is a per-token binary cross-entropy loss

but its not actually a per-token BCE, it is per-sequence BCE. What is per-token is the reward head, since it outputs a logit per token hidden state. Each per-token logit is then `.mean()` and passed through a sigmoid to generate `p(s)`. Then, it is this `p(s)` and the binary 0/1 reward of the whole sequence (not per-token) that is used in the BCE to match the referenced paper's loss in the quote above, screenshot for convenience:
<img width="557" height="141" alt="Image" src="https://github.com/user-attachments/assets/d743351d-0705-4655-950e-7f72b0e4e5e4" />

This leads to the second issue, found in the training script. Notice how in the ss above it will be the BCE on the sigmoid of the mean token logit. In the ORM training script we have:
```python
hidden = self.get_hidden_states(input_ids, attention_mask)
logits = self.head(hidden).squeeze(-1)

loss = None
if labels is not None:
    mask = labels != -100
    if mask.any():
        loss = F.binary_cross_entropy_with_logits(logits[mask], labels[mask].float())
    else:
        loss = logits.sum() * 0
```
Passing the token logits from the reward model (`logits`) to `F.binary_cross_entropy_with_logits` will compute the BCE on the sigmoid of each token logit individually! Aka the issue inside the BCE boils down to the fact that $$\sigma(\frac{1}{T}\sum_t^T w(s_t)) \ne \sigma(w(s_t))$$ where the left-side is the paper's `p(s)` and the right is how we are currently, and incorrectly, computing `p(s)`.

This loss function issue may also explain why the ORM script doesnt converge and hovers just over random guessing of 52% acc. Once the fix is in place we get a much nicer training curve. As a side note, it might be over fitting by learning that if the sequence ends with "Therefore, the answer is" since that is how we construct incorrect sequences, but this is for another time!
<img width="2882" height="982" alt="Image" src="https://github.com/user-attachments/assets/b19f1ded-92d1-443d-a691-c05a48068471" />

**Comment by natolambert** [comment]
This looks amazing thanks so much, will take a deeper look soon. 🙏🩵

**Comment by natolambert** [comment]
Okay I'm pretty happy with #516 to clarify the math / chapter flow. I'm not super urgent to fix the code, but we can follow up on discord and slowly plug away at it. We can leave the issue open for now.

**Comment by natolambert** [comment]
Seems resolved!
